Refresh the HA guide (#24479)
gravitational/teleport: docs/pages/deploy-a-cluster/high-availability.mdx

## Overview

A high-availability Teleport cluster revolves around two pools of redundant
`teleport` processes, one running the Auth Service and one running the Proxy
Service, plus the infrastructure required to support each pool. 

Infrastructure components include:
- A **public Layer 4 load balancer** to direct traffic from users and services

>   to an available `teleport` process.

Reviewer: ```suggestion
  to an available Teleport Proxy Service.
```

![Diagram of a high-availability Teleport
architecture](../../img/deploy-a-cluster/teleport-ha-architecture.png)

## Layer 4 load balancers

High-availability Teleport clusters require two load balancers:
- **Proxy Service load balancer:** A load balancer to receive traffic from
  outside the network where your Teleport cluster is running and forward it to
  an available Proxy Service instance. This load balancer handles TCP traffic
  from users and services in a variety of application-layer protocols.
- **Auth Service load balancer:** A load balancer to forward traffic from a
  Proxy Service instance to an available Auth Service instance. This handles TLS
  traffic to the Auth Service's gRPC endpoint. It can also handle traffic to
  Teleport Kubernetes Service instances running in the Auth Service pool in
  order to enable secure access to your Teleport cluster's infrastructure.

> Neither load balancer must terminate TLS. Both load balancers must transparently

Reviewer: "Neither load balancer must" makes it sound like terminating TLS is optional and the LB doesn't have to, but can, terminate TLS.

Should we be stricter here and say "You must not terminate TLS on either of these load balancers"??

(That said, we technically do support ALBs for most protocols now..)

Reviewer: Paging @greedy52 who can answer this better than I can

Reviewer 2: Afaik:
- Proxy -> Auth LB must not terminate TLS
- Client -> Proxy LB can terminate TLS since 13.0 (but this might come with additional restrictions for kube access)

Author: Thanks! What additional restrictions are these? Is there anything else I should know,  @greedy52?

Reviewer 3: As mentioned, it doesn't support all protocols yet and it also has some performance penalties which we may do some performance improvement in v14. So I think layer 4 LBs should still be recommended way to set up HA.

I was planning to write up a separate page for TLS Routing behind layer 7 LB preview. Maybe I can add a note here to link to that page when it's available?

Author: Sounds good! I'll keep the recommendation the way it is for now and we can add the note when the L7 guide is ready.

Author: I think "Both..must not" is ambiguous, since it can also mean "one can". I've suggested: 

```
Both load balancers must transparently forward the TCP traffic they receive,
without terminating TLS. In other words, these must be Layer 4 load balances,
not Layer 7 (e.g., HTTP) load balancers.
```

| Port | Description |
| - | - |
| <nobr>`3023`</nobr> | SSH port for clients connect to. |
| `3024` | SSH port used to create reverse SSH tunnels from behind-firewall environments. |

> | `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |

Reviewer: Do we specifically tell the user to configure the Teleport proxy to listen on port 443 in this guide?

Would it make more sense to list both the frontend and backend ports here, and have the LB direct incoming traffic on port 443 to port 3080 on the proxy? (Teleport would require additional privileges to bind to 443 as opposed to 3080)

Author: Looks like there's no `web_listen_addr` example here. I'll change the LB instructions!

</TabItem>
</Tabs>

### Configuring the Auth Service load balancer

The Auth Service load balancer must forward traffic to the Auth Service's gRPC
port. In this guide, we are assuming that you have configured the Auth Service
load balancer to forward traffic from port `3025` to port `3025` on an available
Auth Service instance. 

> If you are running your Teleport cluster on Kubernetes, you can configure your

Reviewer: Hmm, this confuses me a bit. Auth never listens on port 3026.

Author: I'm basing this on the `teleport-cluster` chart, which enables the Kubernetes Service on Auth Service pods: https://github.com/gravitational/teleport/blob/64448b1b0ed36643aee0bf471171ba677548f4aa/examples/chart/teleport-cluster/templates/auth/_config.common.tpl#L5-L8

## Teleport instances

Run the Teleport Auth Service and Proxy Service as two scalable groups of

> compute resources, for example, using Kubernetes `Deployment`s or AWS Auto

Reviewer: ```suggestion
compute resources, for example, using Kubernetes Deployments or AWS Auto
```

I don't think we need to render the word deployments in a code block.

```yaml
version: v3
auth_service:
  enabled: false
proxy_service:
  enabled: true

>   public_addr: "mycluster.example.com:3080"

Reviewer 3: ```suggestion
  public_addr: "mycluster.example.com:443"
```
for what load balancer serves.

```yaml
version: v3
auth_service:
  proxy_listener_mode: separate
  enabled: false
proxy_service:
  enabled: true
  listen_addr: 0.0.0.0:3023
  tunnel_listen_addr: 0.0.0.0:3024

>   public_addr: "mycluster.example.com:3080"

Reviewer 3: sorry i think there is one more here:
```suggestion
  public_addr: "mycluster.example.com:443"
```

If you are deploying Teleport on a cluster of virtual machines, remove this line
to run the SSH Service and enable secure access to the host.

### Auth Service pool

#### Open ports

Ensure that, on each Auth Service instance, the following ports are open: 

| Port | Description |
| - | - |
| <nobr>`3025`</nobr> | gRPC port to open to Proxy Service instances.|

> | `3026` | Optional port to open for traffic to the Teleport Kubernetes Service (if deploying on Kubernetes).|

Reviewer: I don't understand this. By this logic, should we also add the desktop service port (3028) because you can deploy desktop service on auth instances as well?

(That's a rhetorical question - I believe the answer is no, I just don't understand why kubernetes_service is getting special treatment here)

Author: As above, this is based on the `teleport-cluster` chart, where Auth Service pods run the Kubernetes Service. My understanding is that Auth Service pods would run the Kubernetes Service to enable access to (at least) Teleport pods running on Kubernetes. Is that accurate, @hugoShaka ?

Reviewer 2: Yeah, it's a bit complex and specific to the Helm chart. We packaged kube access at some point and have to live with this. We also cannot connect kube access to the proxy because we can't establish trust with the proxy without relying on external infra (LE, DNS, LoadBalancer) (discussed [here](https://github.com/gravitational/teleport/issues/19275#issuecomment-1477978729)). I tried to extract this service but failed, and users will be upset if we remove it.

tl;dr I would remove mention of the kube access port in a generic HA guide, this is an implementation detail because we made questionable choices

Author: 👍 I've removed the Kubernetes mentions
